packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -860,84 +1029,81 @@ class TextPainter {
-    // Return early if the current layout information is not outdated, even if
-    // _needsPaint is true (in which case _paragraph will be rebuilt in paint).
-    if (_paragraph != null && minWidth == _lastMinWidth && maxWidth == _lastMaxWidth) {
-      return;
+    assert(!maxWidth.isNaN);
+    assert(!minWidth.isNaN);
+    final InlineSpan? text = this.text;
+    if (text == null) {
+      throw StateError('TextPainter.text must be set to a non-null value before using the TextPainter.');
+    }
+    final TextDirection? textDirection = this.textDirection;
+    if (textDirection == null) {
+      throw StateError('TextPainter.textDirection must be set to a non-null value before using the TextPainter.');
     }
+    final _TextPainterLayoutCache? cachedLayout = _layoutCache;
 
-    if (_rebuildParagraphForPaint || _paragraph == null) {
-      _createParagraph();
+    assert(() {
+      _debugNeedsRelayout = false;
+      return true;
+    }());
+    if (cachedLayout != null && cachedLayout._resizeToFit(minWidth, maxWidth, textWidthBasis)) {
+      return;
     }
-    _lastMinWidth = minWidth;
-    _lastMaxWidth = maxWidth;
-    // A change in layout invalidates the cached caret and line metrics as well.
-    _lineMetricsCache = null;
-    _previousCaretPosition = null;
-    _previousCaretPrototype = null;
-    _layoutParagraph(minWidth, maxWidth);
-    _inlinePlaceholderBoxes = _paragraph!.getBoxesForPlaceholders();
+    // TODO(LongCatIsLooong): remove this when _applyFloatingPointHack is
+    // removed.
+    final double paintOffsetAlignment = _computePaintOffsetFraction(textAlign, textDirection);
+    // Try to avoid laying out the paragraph with maxWidth=double.infinity
+    // when the text is not left-aligned, so we don't have to deal with an
+    // infinite paint offset.
+    final bool adjustMaxWidth = !maxWidth.isFinite && paintOffsetAlignment != 0;
+    final double? adjustedMaxWidth = !adjustMaxWidth ? maxWidth : cachedLayout?.layout.maxIntrinsicLineExtent;
+    final double inputWidth = adjustedMaxWidth ?? maxWidth;
+
+    final ui.Paragraph paragraph = (cachedLayout?.paragraph ?? _createParagraph(text))
+      ..layout(ui.ParagraphConstraints(width: inputWidth));
+
+    final _TextPainterLayoutCache newLayoutCache = _TextPainterLayoutCache(
+      _TextLayout._(paragraph), inputWidth, paintOffsetAlignment, minWidth.floorToDouble(), maxWidth.floorToDouble(), textWidthBasis,
+    );
+    // Call layout again if newLayoutCache had an infinite paint offset.
+    // This is not as expensive as it seems, line breaking is relatively cheap
+    // as compared to shaping.
+    if (adjustedMaxWidth == null && minWidth.isFinite) {
+      assert(maxWidth.isInfinite);
+      final double newInputWidth = newLayoutCache.layout.maxIntrinsicLineExtent;
+      paragraph.layout(ui.ParagraphConstraints(width: newInputWidth));
+      newLayoutCache.inputWidth = newInputWidth;
+    }
+    if (paragraph != cachedLayout?.paragraph) {
```

Comment:
This will only be true when `cachedLayout == null` right? So, `cachedLayout?.paragraph.dispose();` is always a no-op?

Reply:
This is the old paragraph used before the `layout` call so it could be non-null.

Reply:
Ohhh sorry yeah you were right the dispose wasn't necessary.